frameworks/html/src/layers/htmlPictureLayer.ts

```diff
@@ -18,36 +32,171 @@ export class HtmlPictureLayer {
 
     /**
      * Builds source tag bases on user's input
-     * @param param
+     * @param sourceInput
+     * @param autoOptimalBreakpoints
+     * @param relativeWidth
      */
-    buildSourceTag(param: {minWidth?: number, maxWidth?: number, image: CloudinaryImage, sizes?: string}){
+    buildSourceTag(sourceInput: PictureSource, autoOptimalBreakpoints?: boolean, relativeWidth?: number){
         const source = document.createElement('source');
-        source.srcset = param.image.toURL();
+        let srcset: string[] = [];
+
+        if(this.autoOptimalBreakpoints) {
+            this.populateSrcsetUrl(source, sourceInput, srcset, relativeWidth)
+        }
+        this.setSourceAttributes(source, sourceInput, srcset, relativeWidth)
+    }
+
+    /**
+     * Populates srcset based on generated breakpoints and relativeWidth
+     * @param source
+     * @param sourceInput
+     * @param srcset
+     * @param relativeWidth
+     */
+    populateSrcsetUrl(source: HTMLSourceElement, sourceInput: PictureSource, srcset: string[], relativeWidth?: number){
+        let breakPoints = this.getBreakPoints(sourceInput);
+        breakPoints.forEach(breakPoint=>{
+            const width = Math.ceil(breakPoint * relativeWidth);
+            const clone = cloneDeep(sourceInput.image);
+            srcset.push(`${clone.resize(scale(width)).toURL()} ${width}w`);
+        });
+    }
 
-        if(param.sizes) {
-            source.sizes = param.sizes;
+    /**
+     * Returns breakpoints based on min and max width
+     * @param sourceInput
+     */
+    getBreakPoints(sourceInput: PictureSource){
+        if(sourceInput.minWidth > sourceInput.maxWidth){
+            return [sourceInput.minWidth];
         }
-        if(param.maxWidth && param.minWidth){
-            source.media = `(min-width: ${param.minWidth}px) and (max-width: ${param.maxWidth}px)`
-        }else if(param.maxWidth){
-            source.media = `(max-width: ${param.maxWidth}px)`
-        }else if(param.minWidth){
-            source.media = `(min-width: ${param.minWidth}px)`
+        const [physicalMinWidth, physicalMaxWidth] = this.getPhysicalDimension(sourceInput);
+
+        if(physicalMinWidth === physicalMaxWidth){
+            return [physicalMaxWidth];
         }
 
+        return screeWidths.filter((width)=>{
+            return width <= physicalMaxWidth && width >= physicalMinWidth;
+        }).slice(0,SRC_NUMBER)
```

Comment:
This code looks hard for understanding, I would simplify it.

Reply:
It seems pretty straightforward to me.. I updated with comments so should help. Let me know 
if it does:)